### PR TITLE
Add published index endpoint for search reindexing

### PR DESCRIPTION
**Pull request by jessjenkins**
### What

Added PublishedIndex endpoint to zebedee client
- Includes paging parameters (although zebedee currently ignores these)
- Corresponding zebedee PR: https://github.com/ONSdigital/zebedee/pull/551

### How to review

Check code and unit tests

### Who can review

Anyone but me

### CODE REVIEW

**Comment by goofballLogic** [comment]
Do we have standards to review against? Or is it just "look for anything" ?

File: zebedee/client.go [hunk 1]
```diff
@@ -469,6 +469,31 @@ func (c *Client) GetPublishedData(ctx context.Context, uriString string) ([]byte
...
+
+	if params != nil {
+		reqURL = fmt.Sprintf("%s?offset=%d",reqURL, params.Offset)
+		if params.Limit != nil {
+			reqURL = fmt.Sprintf("%s&limit=%d",reqURL, *(params.Limit))
+		}
+	}
+
+	b, _, err := c.get(ctx, "", reqURL)
+
```
Comment (rahulPalliyalil) [inline zebedee/client.go:484]: (suggestion): Extra space, I think the space is not needed from func return and err check

File: zebedee/client_test.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"io/ioutil"
```
Comment (rahulPalliyalil) [inline zebedee/client_test.go:7]: (fyi): just note that ioutil is deprecated from go 1.17 and is moved to the io package. So an update of the go version requires updating this to io.

File: zebedee/client.go [hunk 3]
```diff
@@ -469,6 +469,31 @@ func (c *Client) GetPublishedData(ctx context.Context, uriString string) ([]byte
 	return content, nil
 }
 
+// GetPublishedIndex returns PublishedIndex
+func (c *Client) GetPublishedIndex(ctx context.Context, params *PublishedIndexRequestParams) (PublishedIndex, error) {
+	reqURL := "/publishedindex"
```
Comment (goofballLogic) [inline zebedee/client.go:474]: I notice in the functions above, they form reqURL using `c.createRequestURL` whereas this one is hard coded

File: zebedee/client_test.go [hunk 4]
```diff
@@ -607,3 +608,95 @@ func TestClient_PublishedDataEndpoint(t *testing.T) {
 		})
 	})
 }
+
+func TestClient_PublishedIndexEndpoint(t *testing.T) {
+	ctx := context.Background()
```
Comment (rahulPalliyalil) [inline zebedee/client_test.go:613]: (minor): Can these tests be run in parallel ? t.Parallel()

**Review by rahulPalliyalil — DISMISSED** [review]

File: zebedee/client.go (hunk 3 above)
Comment (jessjenkins) [reply zebedee/client.go:474]: Other functions have dynamic parameters and can take a collectionId. This one only has a simple endpoint so this isn't currently needed and can simply be a string.

File: zebedee/client_test.go (hunk 2 above)
Comment (jessjenkins) [reply zebedee/client_test.go:7]: Thanks. I'll replace with `os.ReadFile(…)` as per deprecation notes in https://pkg.go.dev/io/ioutil#ReadFile

File: zebedee/client_test.go [hunk 5]
```diff
@@ -607,3 +608,95 @@ func TestClient_PublishedDataEndpoint(t *testing.T) {
...
+
+			Convey("then the expected content is returned", func() {
+				So(err, ShouldBeNil)
+				So(m.Count, ShouldEqual, 3)
+				So(m.Items, ShouldNotBeEmpty)
+				So(m.Items, ShouldHaveLength, 3)
+				mItem := m.Items[0]
+				So(mItem.URI, ShouldResemble, "/releases")
+				So(m.Limit, ShouldEqual, 3)
+				So(m.Offset, ShouldEqual, 0)
```
Comment (goofballLogic) [inline zebedee/client_test.go:636]: This will pass even if offset is not specified in the Zebedee response (0 is the default value). Is that the intention?

File: zebedee/client_test.go [hunk 6]
```diff
@@ -607,3 +608,95 @@ func TestClient_PublishedDataEndpoint(t *testing.T) {
...
+				So(m.TotalCount, ShouldEqual, 3)
+			})
+
+			Convey("and client.Do should be called once with the expected parameters", func() {
+				doCalls := httpClient.DoCalls()
+				So(doCalls, ShouldHaveLength, 1)
+				So(doCalls[0].Req.URL.Path, ShouldEqual, path)
+				//p := doCalls[0].Req.FormValue("offset")
```
Comment (goofballLogic) [inline zebedee/client_test.go:644]: some comments left over here?

File: zebedee/client.go [hunk 7]
```diff
@@ -469,6 +469,31 @@ func (c *Client) GetPublishedData(ctx context.Context, uriString string) ([]byte
...
+
+	if params != nil {
+		reqURL = fmt.Sprintf("%s?offset=%d",reqURL, params.Offset)
+		if params.Limit != nil {
+			reqURL = fmt.Sprintf("%s&limit=%d",reqURL, *(params.Limit))
```
Comment (goofballLogic) [inline zebedee/client.go:479]: does it matter that this accepts arbitrarily large numbers (e.g. 9999999999 goes through)

File: zebedee/client_test.go [hunk 8]
```diff
@@ -607,3 +608,95 @@ func TestClient_PublishedDataEndpoint(t *testing.T) {
...
+				doCalls := httpClient.DoCalls()
+				So(doCalls, ShouldHaveLength, 1)
+				So(doCalls[0].Req.URL.Path, ShouldEqual, path)
+				req := doCalls[0].Req
+				So(req.FormValue("offset"), ShouldEqual, "5")
+				So(req.FormValue("limit"), ShouldEqual, "9")
+			})
+		})
+	})
+
```
Comment (goofballLogic) [inline zebedee/client_test.go:678]: Do 404 responses from Zebedee count as errors (in which case no extra test needed), or as empty collections?

File: zebedee/client.go (hunk 7 above)
Comment (goofballLogic) [reply zebedee/client.go:479]: I guess Limit could also be specified as 0 if that matters

File: zebedee/client_test.go (hunk 8 above)
Comment (goofballLogic) [reply zebedee/client_test.go:678]: Zebedee never returns 404 unless the wrong thing was requested

File: zebedee/client.go (hunk 7 above)
Comment (goofballLogic) [reply zebedee/client.go:479]: not checked at this level

File: zebedee/client_test.go (hunk 5 above)
Comment (jessjenkins) [reply zebedee/client_test.go:636]: TBH the paging variables are only there for future compatibility should we implement it. So no need to worry about it right now.

File: zebedee/client_test.go (hunk 8 above)
Comment (jessjenkins) [reply zebedee/client_test.go:678]: No 404 test is needed for the index as it's not specifying a specific resource to return.

**Review by rahulPalliyalil — APPROVED** [review]
